homeassistant/components/s3/backup.py

```diff
@@ -0,0 +1,328 @@
+    hass.data.setdefault(DATA_BACKUP_AGENT_LISTENERS, []).append(listener)
+
+    @callback
+    def remove_listener() -> None:
+        """Remove the listener."""
+        hass.data[DATA_BACKUP_AGENT_LISTENERS].remove(listener)
+        if not hass.data[DATA_BACKUP_AGENT_LISTENERS]:
+            del hass.data[DATA_BACKUP_AGENT_LISTENERS]
+
+    return remove_listener
+
+
+def suggested_filenames(backup: AgentBackup) -> tuple[str, str]:
+    """Return the suggested filenames for the backup and metadata files."""
+    base_name = suggested_filename(backup).rsplit(".", 1)[0]
+    return f"{base_name}.tar", f"{base_name}.metadata.json"
+
+
+class S3BackupAgent(BackupAgent):
+    """Backup agent for the S3 integration."""
+
+    domain = DOMAIN
+
+    def __init__(self, hass: HomeAssistant, entry: S3ConfigEntry) -> None:
+        """Initialize the S3 agent."""
+        super().__init__()
+        self._client = entry.runtime_data
+        self._bucket: str = entry.data[CONF_BUCKET]
+        self.name = entry.title
+        self.unique_id = entry.entry_id
+        self._backup_cache: dict[str, AgentBackup] = {}
+        self._cache_expiration = time()
+
+    @handle_boto_errors
+    async def async_download_backup(
+        self,
+        backup_id: str,
+        **kwargs: Any,
+    ) -> AsyncIterator[bytes]:
+        """Download a backup file.
+
+        :param backup_id: The ID of the backup that was returned in async_list_backups.
+        :return: An async iterator that yields bytes.
+        """
+        backup = await self._find_backup_by_id(backup_id)
+        tar_filename, _ = suggested_filenames(backup)
+
+        response = await self._client.get_object(Bucket=self._bucket, Key=tar_filename)
+        return response["Body"].iter_chunks()
+
+    async def async_upload_backup(
+        self,
+        *,
+        open_stream: Callable[[], Coroutine[Any, Any, AsyncIterator[bytes]]],
+        backup: AgentBackup,
+        **kwargs: Any,
+    ) -> None:
+        """Upload a backup.
+
+        :param open_stream: A function returning an async iterator that yields bytes.
+        :param backup: Metadata about the backup that should be uploaded.
+        """
+        tar_filename, metadata_filename = suggested_filenames(backup)
+
+        try:
+            if backup.size < MULTIPART_MIN_PART_SIZE_BYTES:
+                await self._upload_simple(tar_filename, open_stream)
+            else:
+                await self._upload_multipart(tar_filename, open_stream)
+
+            # Upload the metadata file
+            metadata_content = json.dumps(backup.as_dict())
+            await self._client.put_object(
+                Bucket=self._bucket,
+                Key=metadata_filename,
+                Body=metadata_content,
+            )
+
+            # Reset cache after successful upload
+            self._cache_expiration = time()
+
+        except BotoCoreError as err:
+            raise BackupAgentError("Failed to upload backup") from err
```

Comment:
```suggestion
        except BotoCoreError as err:
            raise BackupAgentError("Failed to upload backup") from err
       else:
            # Reset cache after successful upload
            self._cache_expiration = time()
```
only have code in try that can fail